DynamoRIO/dynamorio
# i#5946: Check whether dynamic info is set in os_module_data_t before using it

This PR tries to fix #5946 by performing extra check for dynamic info was initialized in os_module_data_t during module load.

The error is triggered when loaded ELF module have a specific program sections order: `.dynstr`  should be placed in the end. In such case it wouldn't be mapped at the moment of os_module_data_t initialization.
If DR-client would try to use module data later (via dr_get_proc_address for example), DR will crash.

_How to reproduce for testing:_
1) You need a target program with dynamic linking to the shared library (example: )
2) Shared library ELF should have late .dynstr section. This could be achieved by using `patchelf` utility: after a compilation of libtest.so object you need to run `patchelf --set-rpath '$ORIGIN' libtest.so`. To check program sections displacement run `readelf -l libtest.so`
3) The crash is triggered when DR client uses uninitialized os_module_data_t. The easiest way is to call dr_get_proc_address() in module_load_event handler.

Test example with compiled binaries and source code: [example.tar.gz](https://github.com/DynamoRIO/dynamorio/files/11200545/example.tar.gz).
Tested on ubuntu20.04/ubuntu22.04 with clang-14 and patchelf >=0.10
